### PR TITLE
[WIP, enhancement] enable array_api return values from ```from_table```

### PR DESCRIPTION
**Pull request by icfaust**
## Description

NOTE: The initial support version for ```from_table``` is not general enough and highly coded for sycl usm inputs.  This changes the interface to mirror the type (and possibly queue) of the data.  Getting this integrated will begin the fight against the technical debt introduced with ```use_raw_inputs```. The core of this PR is simple, the changes needed associated with the technical debt will be brutal.

An open question exists for type persistence in incremental algorithms, which do not follow use_raw_inputs in the return types (probably for this reason).  Some sort of solution must be found and implemented, probably as a follow-up to this PR.  It is likely that a conversion function must be stored in the instance along with the queue for returning the proper value.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.
- [ ] I have provided justification why quality metrics have changed or why changes are not expected.
- [ ] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: onedal/datatypes/_data_conversion.py [hunk 1]
```diff
@@ -34,84 +34,108 @@ def to_table(*args, queue=None):
...
+    dpctl/dpnp usm_ndarrays, array API standard arrays, and scalars. Tables
+    will use pointers to the original array data. Scalars and non-contiguous
+    arrays will be copies. Arrays may be modified in-place by oneDAL during
+    computation. This works for data located on CPU and SYCL-enabled Intel GPUs.
+    Each array may only be of a single datatype (i.e. each must be homogeneous).
 
     Parameters
     ----------
-    *args : {scalar, numpy array, sycl_usm_ndarray, csr_matrix, or csr_array}
+    *args : scalar, numpy array, sycl_usm_ndarray, array API standard array,
```
Comment (david-cortes-intel) [inline onedal/datatypes/_data_conversion.py:45]: Is `sycl_usm_ndarray` this specific class? https://intelpython.github.io/dpctl/latest/api_reference/dpctl/generated/dpctl.tensor.usm_ndarray.html

File: onedal/datatypes/_data_conversion.py [hunk 2]
```diff
@@ -34,84 +34,108 @@ def to_table(*args, queue=None):
...
+        csr_matrix, or csr_array
         arg1, arg2... The arrays should be given as arguments.
 
+    queue : SyclQueue or None, default=None
+        A dpctl or oneDAL backend python representation of a SYCL Queue or None
+
     Returns
     -------
-    tables: {oneDAL homogeneous tables}
+    tables: oneDAL homogeneous tables
```
Comment (david-cortes-intel) [inline onedal/datatypes/_data_conversion.py:54]: Format here would need to be:
```
name_of_returned_variable : type_of_returned variable
    Description about it.
```

**Comment by icfaust** [comment]
This has implications about array-api-compat. I need to rethink this strategy.

**Comment by icfaust** [comment]
The situation is as follows. I didn't look thoroughly enough into the reasons behind array-api-compat and its inclusion into sklearn.  Even if we fully support array_api inputs, it will not mean we will support pytorch out of the box.  Pytorch support via array api comes via getting the namespace generating from array-api-compat, meaning manually searching for __array_namespace__ won't work (and the reason why sklearn has ```get_namespace``` to compensate for that.  Thus either we need to re-invent ```get_namespace``` in onedal, acquire array namespaces from sklearn and pass them around like we do for queues (not advised) having used sklearn's ```get_namespace``` or think of another way of doing it.  The logic is insufficient for pytorch support.

**Comment by icfaust** [comment]
I'm going to have to modify the injestion scripts (```to_table```) for non-contiguous inputs and dtype changes which we have a special case trigger in sycl-usm, which we will also have to add for pytorch...

**Comment by icfaust** [comment]
> The situation is as follows. I didn't look thoroughly enough into the reasons behind array-api-compat and its inclusion into sklearn. Even if we fully support array_api inputs, it will not mean we will support pytorch out of the box. Pytorch support via array api comes via getting the namespace generating from array-api-compat, meaning manually searching for **array_namespace** won't work (and the reason why sklearn has `get_namespace` to compensate for that. Thus either we need to re-invent `get_namespace` in onedal, acquire array namespaces from sklearn and pass them around like we do for queues (not advised) having used sklearn's `get_namespace` or think of another way of doing it. The logic is insufficient for pytorch support.

I ended up doing a final catch for types with a lazy load of ```array_api_compat``` which may or may not be installed with sklearn.  If someone tries to use sklearnex with other array types (like pytorch), then they will have installed compat anyway to interface with sklearn.  If someone uses the oneDAL estimators directly, then they will need the warning for pytorch in the case that the sklearn install doesn't exist.

File: onedal/linear_model/logistic_regression.py [hunk 3]
```diff
@@ -209,20 +209,15 @@ def _infer(self, X):
...
-            xp=xp,
-        )
+        y = from_table(result.responses, like=X)
         y = xp.take(xp.asarray(self.classes_), xp.reshape(y, (-1,)), axis=0)
         return y
 
     def _predict_proba(self, X):
         result = result = self._infer(X)
         sua_iface, xp, _ = _get_sycl_namespace(X)
         queue = QM.get_global_queue()
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/linear_model/logistic_regression.py:219]: lol, oh copilot.

File: onedal/datatypes/_data_conversion.py [hunk 4]
```diff
@@ -61,50 +65,93 @@ def to_table(*args, queue=None):
...
+        When set to None, will return numpy arrays or scipy csr arrays.
+
+    Returns
+    -------
+    func : callable
+        A function which takes in a single table input and returns an array.
+    """
+    if isinstance(array, np.ndarray) or array is None or sp.issparse(array):
+        func = backend.from_table
+    elif hasattr(array, "__sycl_usm_array_interface__"):
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/datatypes/_data_conversion.py:89]: If some data ever gets that far to fail in that way, I will shake their hand.

File: onedal/datatypes/_data_conversion.py [hunk 5]
```diff
@@ -37,16 +38,19 @@ def to_table(*args, queue=None):
...
-    *args : {scalar, numpy array, sycl_usm_ndarray, csr_matrix, or csr_array}
+    *args : scalar, numpy array, sycl_usm_ndarray, csr_matrix, or csr_array
         Arguments to be individually converted to oneDAL tables.
 
     queue : SyclQueue or None, default=None
         SYCL Queue object to be associated with the oneDAL tables. Default
         value None causes no change in data location or queue.
 
+    queue : SyclQueue or None, default=None
+        A dpctl or oneDAL backend python representation of a SYCL Queue or None.
```
Comment (david-cortes-intel) [inline onedal/datatypes/_data_conversion.py]: Argument has two tocs.

**Comment by david-cortes-intel** [comment]
@icfaust What happens if the 'X' and 'y' inputs are of different types, or if one is on device and another on host?

File: onedal/datatypes/_data_conversion.py (hunk 5 above)
Comment (icfaust) [reply onedal/datatypes/_data_conversion.py]: shoot. good catch, i assume there was a bad main merge

**Comment by icfaust** [comment]
> @icfaust What happens if the 'X' and 'y' inputs are of different types, or if one is on device and another on host?

Good question, this goes to some underlying open questions about the onedal module estimators and their direct use.  In sklearnex there are guards for this associated with ```dispatch()```  which can be seen/ are updated in #2569 (see ```_sycl_queue_manager.from_data```).  However, using the onedal estimators directly (which is the case for spmd as that code is currently completely unregulated), disregards finiteness, shape, input type, verification of the input queue, etc. In sklearnex these are handled by ```validate_data``` with some additional constraints as previously described.

With respect to the second question: if return data is on GPU, it will try to return GPU data to the array constructor.  If input data was with a cpu sycl queue, it will attempt to copy the cpu queue from the input data (only pertinent for  array types with ```__sycl_usm_ndarray_interface__``` attributes.

Hope this helps.

**Comment by david-cortes-intel** [comment]
@icfaust Could you please add tests having (a) 'X' as numpy and 'y' as GPU array; (b) 'X' as GPU array and 'y' as numpy; (c) 'X' as sycl host array on torch and 'y' as sycl device array on torch.

I'm not sure if this is the most suitable PR though.

**Comment by icfaust** [comment]
> @icfaust Could you please add tests having (a) 'X' as numpy and 'y' as GPU array; (b) 'X' as GPU array and 'y' as numpy; (c) 'X' as sycl host array on torch and 'y' as sycl device array on torch.
> 
> I'm not sure if this is the most suitable PR though.

Thank you for the suggestion, I will add testing to that effect in #2569 

EDIT: this mixing of types will be only problematic for ```array_api_dispatch=True```, but should be caught via the proper use of ```get_namespace```. Likely a new design rule must be added that ```get_namespace``` must be run for all args, as that occurs independently for each estimator and method (at the moment). @david-cortes-intel

File: onedal/datatypes/_data_conversion.py [hunk 6]
```diff
@@ -37,16 +38,19 @@ def to_table(*args, queue=None):
...
+
     Returns
     -------
-    tables: {oneDAL homogeneous tables}
+    tables : oneDAL homogeneous tables
```
Comment (Vika-F) [inline onedal/datatypes/_data_conversion.py]: Shoulld it be?
```suggestion
    tables : oneDAL homogeneous tables for scalar, numpy array, sycl_usm_ndarray; oneDAL csr_table for csr_matrix, csr_array
```

File: onedal/covariance/covariance.py [hunk 7]
```diff
@@ -102,29 +102,27 @@ def fit(self, X, y=None, queue=None):
...
         use_raw_input = _get_config()["use_raw_input"] is True
-        sua_iface, _, _ = _get_sycl_namespace(X)
+        sua_iface, xp, _ = _get_sycl_namespace(X)
         if use_raw_input and sua_iface:
             queue = X.sycl_queue
 
         if not use_raw_input:
             X = _check_array(X, dtype=[np.float64, np.float32])
-        X = to_table(X, queue=queue)
+        X_t = to_table(X, queue=queue)
```
Comment (Vika-F) [inline onedal/covariance/covariance.py]: It would be better to replace `X_t` with `X_table` as `X_t` has a misleading connotation related to X transposed.

File: onedal/ensemble/forest.py [hunk 8]
```diff
@@ -390,14 +390,14 @@ def _predict(self, X, hparams=None):
 
         model = self._onedal_model
         queue = QM.get_global_queue()
-        X = to_table(X, queue=queue)
-        params = self._get_onedal_params(X)
+        X_t = to_table(X, queue=queue)
```
Comment (Vika-F) [inline onedal/ensemble/forest.py]: Same with `X_t` here.

File: onedal/covariance/covariance.py (hunk 7 above)
Comment (icfaust) [reply onedal/covariance/covariance.py]: Thats a valid point, will switch.

File: onedal/cluster/dbscan.py [hunk 9]
```diff
@@ -78,15 +78,17 @@ def fit(self, X, y=None, sample_weight=None, queue=None):
         params = self._get_onedal_params(X_table.dtype)
         result = self.compute(params, X_table, sample_weight_table)
 
-        self.labels_ = from_table(result.responses, sycl_queue=queue).ravel()
+        self.labels_ = xp.squeeze(from_table(result.responses, like=X))
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/cluster/dbscan.py:81]: See line 68

File: onedal/ensemble/forest.py (hunk 8 above)
Comment (icfaust) [reply onedal/ensemble/forest.py]: swapped.

File: onedal/datatypes/_data_conversion.py (hunk 6 above)
Comment (icfaust) [reply onedal/datatypes/_data_conversion.py]: Good call, changed to reflect that.

**Review by Vika-F — APPROVED** [review]
The failures in CI (WindowsCondaEnv Python3.12_Sklearn1.6) are not related to these changes, as they appear in main as well.
Let's wait for a private CI and merge.
